datavault/backend.py

```diff
@@ -216,7 +439,378 @@ def hasMore(self, pos):
+            column_types.append(shape_tag + base_type + unit_tag + comment)
+        type_tag = '*({})'.format(','.join(column_types))
+        return type_tag
+
+    def getTransposeType(self):
+        column_type = []
+        for col in self.getIndependents() + self.getDependents():
+            base_type = col.datatype
+            if base_type in ['v', 'c']:
+                unit_tag = '[{}]'.format(col.unit)
+            else:
+                unit_tag = ''
+            if len(col.shape) > 1:
+                shape_tag = '*{}'.format(len(col.shape) + 1)
+                comment = util.braced('N,' + ','.join(str(s) for s in col.shape))
+            elif col.shape[0] > 1:
+                shape_tag = '*2'
+                comment = util.braced('N,' + str(col.shape[0]))
+            else:
+                shape_tag = '*'
+                comment = ''
+            column_type.append(shape_tag + base_type + unit_tag + comment)
+        type_tag = '({})'.format(','.join(column_type))
+        return type_tag
+
+    def addParam(self, name, data):
+        keyname = 'Param.{}'.format(name)
+        value = labrad_urlencode(data)
+        self.dataset.attrs[keyname] = value
+
+    def getParameter(self, name, case_sensitive=True):
+        """Get a parameter from the dataset."""
+        keyname = 'Param.{}'.format(name)
+        if case_sensitive:
+            if keyname in self.dataset.attrs:
+                return labrad_urldecode(self.dataset.attrs[keyname])
+        else:
+            for k in self.dataset.attrs:
+                if k.lower == keyname.lower:
+                    return labrad_urldecode(self.dataset.attrs[k])
+        raise errors.BadParameterError(name)
+
+    def getParamNames(self):
+        """Get the names of all dataset parameters.
+
+        Parameter names in the HDF5 file are prefixed with 'Param.' to avoid
+        conflicts with the other metadata.
+        """
+        names = [str(k[6:]) for k in self.dataset.attrs if k.startswith('Param.')]
+        return names
+
+    def addComment(self, user, comment):
+        """Add a comment to the dataset."""
+        t = time.time()
+        new_comment = np.array([(t, user, comment)], dtype=self.comment_type)
+        old_comments = self.dataset.attrs['Comments']
+        data = np.hstack((old_comments, new_comment))
+        self.dataset.attrs.create('Comments', data, dtype=self.comment_type)
+
+    def getComments(self, limit, start):
+        """Get comments in [(datetime, username, comment), ...] format."""
+        if limit is None:
+            raw_comments = self.dataset.attrs['Comments'][start:]
+        else:
+            raw_comments = self.dataset.attrs['Comments'][start:start+limit]
+        comments = [(datetime.datetime.fromtimestamp(c[0]), str(c[1]), str(c[2])) for c in raw_comments]
+        return comments, start+len(comments)
+
+    def numComments(self):
+        return len(self.dataset.attrs['Comments'])
+
+class ExtendedHDF5Data(HDF5MetaData):
```

Comment:
ExtendedHDF5Data and SimpleHDF5Data don't have a .data attribute whereas CsvListData and CsvNumpyData do - is that intentional?